src/key_values/value/any.rs

```diff
@@ -0,0 +1,97 @@
+use std::{fmt, mem};
+use std::marker::PhantomData;
+
+use super::{ToValue, KeyValueError};
+use super::backend::Backend;
+
+/// A function for converting some type `T` into a [`Value`](struct.Value.html).
+pub type FromAnyFn<T> = fn(FromAny, &T) -> Result<(), KeyValueError>;
+
+/// A helper for converting any type into a [`Value`](struct.Value.html).
+pub struct FromAny<'a>(&'a mut Backend);
+
+impl<'a> FromAny<'a> {
```

Comment:
This type looks like "&mut dyn Backend, but with extra helpers". It sounds like we'll want Backend in the public API eventually, so consider putting the helpers directly on dyn Backend instead:

```rust
impl dyn Backend + '_ {
    pub(crate) fn value<T>(&mut self, v: T) -> Result<(), KeyValueError>
    where
        T: ToValue,
    {...}

    pub fn debug<T>(&mut self, v: T) -> Result<(), KeyValueError>
    where
        T: Debug,
    {...}
}
```

We can seal the Backend trait for now if we'd like to be free to add mandatory methods.

Reply:
The idea behind the `FromAny` type (and the rest of the `any` API here) is to try really hard not to make `Value` look like a serialization contract. The reason is mostly philosophical. The exercise of building out the `sval` framework convinced me that building a serialization API is tricky, and design decisions can seem arbitrary but have significant impacts on usability and interoperability (`sval` has some really wacky code to integrate with `serde`), so `log` should avoid owning that story and defer to existing libraries.

As an example, the way I imagine we'd add `serde` support to `Value` would be:

```rust
impl<'v> Value<'v> {
    /// Capture a type that implements `serde::Serialize` as a `Value.
    pub fn from_serde(v: &'v impl Serialize) -> Self {
        Self::from_any(v, |from, v| from.serde(v)) 
    }
}

impl<'v> Serialize for Value<'v> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        //  an implementation of the internal `Backend` trait
        let mut backend = SerdeBackend::new(serializer);
        self.visit(&mut backend)?;
        
        backend.into_result()
    }
}

impl<'b> FromAny<'b> {
    pub fn serde(self, v: impl Serialize) -> Result<(), KeyValuesError> {
        self.0.serde(v)
    }
}

trait Backend {
    /// Visit a `serde::Serialize`. All backends have to support visiting a `Serialize`.
    fn serde(&mut self, v: &erased_serde::Serialize) -> Result<(), KeyValuesError>;
}
```

Internally, we can't get away with not having some contract for frameworks to conform to, but we  can avoid ever exposing it directly. Instead, we can implement traits like `std::fmt::Debug`, `serde::Serialize`, and`sval::Value` on `log::Value` as the way you can inspect the underlying structure instead of implementing a trait.

But as someone with a lot of experience working on serialization APIs I'd welcome any input you've got here!